producer.go

```diff
@@ -71,6 +86,48 @@ func (p *segmentProducer) newWriter(cfg Config) error {
 	return nil
 }
 
+// onMessageDelivery is a callback function for kafka-go's writer
+// which is calling the ack functions for the messages which were
+// (successfully or unsuccessfully) delivered to Kafka.
+func (p *segmentProducer) onMessageDelivery(messages []kafka.Message, err error) {
+	if len(messages) == 0 {
+		return
+	}
+	for _, m := range messages {
+		// accessed also when in Send(), when messages need to be sent
+		p.m.Lock()
+		ackFunc, ok := p.ackFuncs[p.getID(m)]
+		delete(p.ackFuncs, p.getID(m))
+		p.m.Unlock()
+
+		if !ok {
+			sdk.Logger(context.Background()).
+				Error().
+				Msgf("no ack function for %v registered", p.getID(m))
+			// Either Conduit didn't send an AckFunc (pretty unlikely)
+			// or this connector got an AckFunc, but didn't store it.
+			// Either way, without this message being acknowledged,
+			// the source record cannot be ack'd either, which means
+			// that no position (progress) past this will be persisted.
+			// That implies that after a restart, everything after this position
+			// will be redone anyway.
+			// By panicking, we make sure there's no wrong impression of everything being just fine.
+			panic(fmt.Errorf("ack func for %v not registered", p.getID(m)))
```

Comment:
I thought about this long and hard and came to the conclusion I don't have a good alternative that won't overly complicate the solution. For the record, I feel queasy about the panic, even though it's unlikely to happen, because if the plugin is running in built-in mode it will bring down the whole Conduit process. And given what I've seen in connectors from other developers, I see that they take a lot of ideas from our in-house connectors, but I wouldn't want them to think panicking in a connector is an acceptable thing to do.

It's becoming more and more clear we need to change the Destination interface, the way we currently expose async writes is hard to implement nicely and contains one too many foot guns.

Reply:
I don't feel good when panicking either.: ) But given everything else, and that we want to simplify the Destination interface, I believe we can live with it for some time.